moshi/src/test/java/com/squareup/moshi/JsonReaderTest.java

```diff
@@ -1433,6 +1434,24 @@ public void nextSourceNotConsumed() throws IOException {
     assertThat(reader.nextString()).isEqualTo("d");
   }
 
+  @Test
+  public void tags() throws IOException {
+    JsonReader reader = newReader("{}");
+    assertThat(reader.getTag(String.class)).isNull();
+    assertThat(reader.getTag(Integer.class)).isNull();
```

Comment:
One extra constraint – we should require that the tag value is an instance of the tag’s type. In practice you’d create a type for ever tag you set.

Reply:
I've added the constraint in the test, as well as adding an exception in the `setTag` method in case someone attempts to assign a wrong value.